lib/api/objectPutLegalHold.js

```diff
@@ -0,0 +1,116 @@
+const async = require('async');
+const { errors, s3middleware } = require('arsenal');
+
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const { decodeVersionId, getVersionIdResHeader } =
+  require('./apiUtils/object/versioning');
+const getReplicationInfo = require('./apiUtils/object/getReplicationInfo');
+const metadata = require('../metadata/wrapper');
+const { metadataValidateBucketAndObj } = require('../metadata/metadataUtils');
+const { pushMetric } = require('../utapi/utilities');
+
+const { parseLegalHoldXml } = s3middleware.objectLegalHold;
+
+const REPLICATION_ACTION = 'PUT_LEGAL_HOLD';
+
+/**
+ * Object Put Legal Hold - Sets legal hold status of object
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function objectPutLegalHold(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'objectPutLegalHold' });
+
+    const { bucketName, objectKey } = request;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return process.nextTick(() => callback(decodedVidResult));
+    }
+    const versionId = decodedVidResult;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        objectKey,
+        requestType: 'objectPutLegalHold',
+        versionId,
+    };
+
+    return async.waterfall([
```

Comment:
Same as above, async.waterfall returns undefined so just not returning has the same affect.

Reply:
You're right about this and `return` was redundant, so initially I made the change but just reverted back due to linter warning failing CI for `consistent-return` "Expected to return a value at the end of this function". (It's interesting as there're bunch of functions with `async.waterfall` without `return` though)

Reply:
We might want to take a look at the linter settings in more detail, more than a few of my comments have been struck down to "linter caused an error" for, of course IMO, no reason. This will only increase as we move to more recent javascript standards like async/await and generators.